### ISSUE TITLE
Change global `pytest` logging level to `INFO` 

### ISSUE THREAD

**Issue by jacobcook1995** [ImperialCollegeLondon/virtual_rainforest#190]
**Is your feature request related to a problem? Please describe.**
As @alexdewar pointed out we currently test for a lot of `DEBUG` level logging output. I can see the argument against doing this as the `DEBUG` logging level is basically used to hide information that would otherwise clutter up the logging output, which makes the expected log tests pretty long and convoluted. And also `DEBUG` logging messages are only likely to be of interest to us as developers.

**Describe the solution you'd like**
We could set to the `pytest` logging to `INFO` so that `DEBUG` messages do not end up in the `caplog` fixture. We would then have to remove `DEBUG` messages from our expected logs for tests.

**Describe alternatives you've considered**
The `pytest` logging level does have to be set globally otherwise you get inconsistent behaviour when `pytest` is run on the whole repo vs run on individual files. That means the only real options are to set the level to `DEBUG`, `INFO`, `WARNING`, or `ERROR`. I feel like `INFO` is the most sensible unless others disagree.

**Comment by davidorme** [comment]
I don't think this is a bad idea but I think this change is part of a bigger issue:

* I don't think we have consistent usage of what DEBUG and INFO logger levels mean yet. I suspect that is something we are going to want to review once we have some actual logs, but existing tests that check DEBUG log messages may indicate that we to increase the message level rather than mute that message.

* Even with consistent levels, I think that some tests will and should intentionally check functionality that emits `DEBUG` messages. This isn't a problem - even if we have a global setting, we can still adjust the captured logging level **within** tests. Tests that do so probably have to restore the setting? I haven't checked if that persists but it seems likely. If it does, I think we should create a test decorator to do this - set the level, run the test, restore the level:

   ```python
   @capture_debug  # or maybe @caplog_level("DEBUG") if there are other levels to consider
   def test_the_debug_messages:
       pass
    ```

* Another thing here is that we may want a verbose mode for Virtual Rainforest. We might be able to do this with DEBUG messages basically being the verbose mode logging, so that just INFO and above are reported in standard mode. But, if we want to keep DEBUG for genuine developer debugging, we might want a custom VERBOSE logger level between DEBUG and INFO. Again, I think this is something we will have a better feel for once we have some logs.

**Comment by alexdewar** [comment]
> I don't think this is a bad idea but I think this change is part of a bigger issue:
> 
> * I don't think we have consistent usage of what DEBUG and INFO logger levels mean yet. I suspect that is something we are going to want to review once we have some actual logs, but existing tests that check DEBUG log messages may indicate that we to increase the message level rather than mute that message.

Agreed. Some of your debugs should probably be turned into infos.

> * Even with consistent levels, I think that some tests will and should intentionally check functionality that emits `DEBUG` messages. This isn't a problem - even if we have a global setting, we can still adjust the captured logging level **within** tests. Tests that do so probably have to restore the setting? I haven't checked if that persists but it seems likely. If it does, I think we should create a test decorator to do this - set the level, run the test, restore the level:
>   ```python
>   @capture_debug  # or maybe @caplog_level("DEBUG") if there are other levels to consider
>   def test_the_debug_messages:
>       pass
>   ```

I'm really not sure about this. You can't test for absolutely every single thing your code does, even if you wanted to, so you do have to be a bit selective. Do you really want to promise your users that output at the debug level won't change? Does anyone need this guarantee? It just seems like it will clutter up your tests needlessly.

I *suspect* that for some of the tests, the debug output might be being used as a proxy to check whether a given code path has been taken. For those cases, you want a more reliable test (e.g. by mocking a function that should be called).

> * Another thing here is that we may want a verbose mode for Virtual Rainforest. We might be able to do this with DEBUG messages basically being the verbose mode logging, so that just INFO and above are reported in standard mode. But, if we want to keep DEBUG for genuine developer debugging, we might want a custom VERBOSE logger level between DEBUG and INFO. Again, I think this is something we will have a better feel for once we have some logs.

I don't think you really need any more logging levels. You've already got:

- `INFO`: Output that everyone, including users, cares about
- `DEBUG`: Output that only VR devs care about
- `VERBOSE`: The nuclear option for when VR devs want to see every little thing that is happening

What else could you need?

At some point you *may* want to consider having different loggers for different models, so that you can look at the debug output for just one of them, but this is definitely not something to do just yet.

> 
> My immediate feeling is that we park this for now and then have a systematic review once we've got models running and emitting logs? Otherwise, I can see us setting this, deleting all the DEBUG log entries in tests and then having to restore them all when we review logging.

Agreed. I don't think you should go ripping out working tests for the sake of it. But I think this issue is something worth considering in the longer run. If you have to add ~5 lines of code to your tests for every debug statement it's just going to get messy, and for little gain AFAICS.

**Comment by davidorme** [comment]
@alexdewar We don't have `VERBOSE` - that isn't one of the `logging` levels. I was thinking of a hierarchy:

* `INFO`: Output that tells the user what the simulation is doing.
* `VERBOSE`: Output that tells the user what the simulation is doing under the hood.
* `DEBUG`: Stuff that VR devs care about

Having said that, the `logging` module advice on custom levels is [don't](https://docs.python.org/3/howto/logging.html#custom-levels) and we could just use `DEBUG` as `VERBOSE`.

On the per-test level - if we do have a function that emits stable DEBUG messages, then testing the logged behaviour is as expected for specific low-level tests of that code seems fine to me. We'd want to suppress those DEBUG logs for other tests that simply used the function. I think mocking is great, but often seems to end up being really opaque.

**Comment by alexdewar** [comment]
So we don't. Ha! I just assumed Python was the same as everyone else in that regard... Generally, debug comes before verbose though, as I said, so it would be confusing to have it the other way round.

I agree with the docs though that custom logging levels are a bad idea. You'll just make things unnecessarily confusing if you break with the conventions that basically all other Python programs use.

And as I said, I don't think you need to guarantee that debug statements are consistent between versions. Mocking might be a bit ugly sometimes, but it's the right way to check that a function is *actually* called, rather than just assuming that the presence of console output means that the code has done what you expect. It's up to you guys though. I just want my disagreement noted :wink:
